### ISSUE TITLE
Bluetooth not shown in Peripherals (Monterey beta 5 / Hackintool 3.6.2)

### ISSUE THREAD

**Issue by perez987** [headkaze/Hackintool#147]
Good afternoon. Thanks for your great app.
In Monterey beta 5, Hackintool 3.6.2 doesn't display Bluetooth in Peripherals tab although it's well detected and well shown in System Profiler. My BT hardware is Fenvi T-919. It works OOTB in Monterey. It seems to be a cosmetic issue.

<img width="757" alt="bt1" src="https://user-images.githubusercontent.com/50429399/129769735-b60fdf3b-dc6a-47a0-a198-8c2c8f860bcb.png">
<img width="652" alt="bt2" src="https://user-images.githubusercontent.com/50429399/129769737-a455270e-e746-4471-93fe-8c25ab5bd461.png">

Note: in Big Sur it's working fine.

**Comment by headkaze** [comment]
Can you post your IOReg dump from [IORegistryExplorer](https://olarila.com/files/Utils/IORegistryExplorer.app.zip)

**Comment by perez987** [comment]
Sorry for the delay, I've been busy. Here is my ioreg. Note that it's from Monterey. Bluetooth system seems to work fine in both Big Sur and Monterey. In Big Sur Hackintool BT device is well shown in peripherals.
[GRANJA1.ioreg.zip](https://github.com/headkaze/Hackintool/files/7112048/GRANJA1.ioreg.zip)

**Comment by feartech** [comment]
> Sorry for the delay, I've been busy. Here is my ioreg. Note that it's from Monterey. Bluetooth system seems to work fine in both Big Sur and Monterey. In Big Sur Hackintool BT device is well shown in peripherals. [GRANJA1.ioreg.zip](https://github.com/headkaze/Hackintool/files/7112048/GRANJA1.ioreg.zip)

ahh! have the same issue with same card, thought it was something i was doing wrong :)

**Comment by perez987** [comment]
@feartech 
In the meantime, I have met more users who experience the same thing, it seems only a cosmetic problem related to the changes that Bluetooth has in Monterey, probably a future update of Hackintool will solve it.

**Comment by headkaze** [comment]
Can you guys let me know if 3.6.4 fixes this issue

**Comment by feartech** [comment]
> Can you guys let me know if 3.6.4 fixes this issue

still doesn't show up in the bluetooth section, only the section above (wifi/ethernet)

**Comment by headkaze** [comment]
@williambj1 any ideas why it's not working in this circumstance?

**Comment by perez987** [comment]
> > Can you guys let me know if 3.6.4 fixes this issue
> 
> still doesn't show up in the bluetooth section, only the section above (wifi/ethernet)

Me too, it still doesn't show (only Monterey, Big Sur is ok).

**Comment by williambj1** [comment]
@feartech @perez987 Does the bluetooth device show up in Hackintool's USB section?

**Comment by feartech** [comment]
> @feartech @perez987 Does the bluetooth device show up in Hackintool's USB section?

yes, it shows up there fine - BRCM20702 Hub

**Comment by perez987** [comment]
I have noticed that in a real iMac, with Monterey, BT does not appear in Peripherals tab either, as in the hack. 
Both machines show BT device in USB tab.

**Comment by headkaze** [comment]
Can you guys please try 3.6.5

**Comment by perez987** [comment]
It still doesn't show in 3.6.5
<img width="688" alt="3 6 5" src="https://user-images.githubusercontent.com/50429399/141693508-9a5c5514-05ca-4ad2-811d-4c735ed2c846.png">
.

**Comment by headkaze** [comment]
Can someone with Monterey take a look at [getIORegBluetoothArray](https://github.com/headkaze/Hackintool/blob/master/Hackintool/IORegTools.m#L1174) in Xcode? Based on @perez987's IORegistry iterating over all IOUSBDevice devices and if the "USB Product String" property is "Bluetooth USB Host Controller" the parent IOUSBDevice should be the Bluetooth device.

**Comment by perez987** [comment]
> Can someone with Monterey take a look at [updateBluetoothDevices](https://github.com/headkaze/Hackintool/blob/master/Hackintool/AppDelegate.m#L2044) and [getIORegBluetoothArray](https://github.com/headkaze/Hackintool/blob/master/Hackintool/IORegTools.m#L1174) in Xcode? Based on @perez987's IORegistry iterating over all IOUSBDevice devices and if the kUSBProductString ("USB Product String") property is "Bluetooth USB Host Controller" the parent IOUSBDevice should be the Bluetooth device.

Sorry but my programming knowledge is far below yours, I can read the methods you say but I am not able to get any valid conclusions.
The only thing I can tell you is what you already know, bluetooth device is not in peripherals tab although it works correctly. And it only happens in Monterey, not in Big Sur.
Otherwise, Hackintool seems to work fine.
Tell me if I can help with something else.

**Comment by headkaze** [comment]
@williambj1 any chance you can check out these new routines in Xcode and see why they're not discovering the BT device?

**Comment by headkaze** [comment]
I just got Monterey installed successfully but BluetoothFixup is causing a KP so I'll take a look later tonight.

**Comment by perez987** [comment]
It's a shame because my Fenvi T919 works fine in Monterey and I would like to help more.

**Comment by williambj1** [comment]
> @williambj1 any chance you can check out these new routines in Xcode and see why they're not discovering the BT device?

@headkaze I couldn't spot the problem in the code based on my very limited knowledge for iterating through IOReg. I only have Intel Bluetooth devices and (unlike BRCM cards) they don't have their own hub.

I'm thinking about taking the same approach bluetoothd in macOS Monterey uses to find bluetooth devices. Would require rewriting some reverse engineered pseudo code.

**Comment by headkaze** [comment]
@williambj1 can you post your IOReg? I believe I've got a working solution now but would like to double check it will work for Intel devices.

**Comment by williambj1** [comment]
[IntelBT.ioreg.zip](https://github.com/headkaze/Hackintool/files/7562934/IntelBT.ioreg.zip)

@headkaze Here's the dump from IORegistryExplorer v3

**Comment by headkaze** [comment]
@williambj1 are you on Monterey?

**Comment by williambj1** [comment]
> @williambj1 are you on Monterey?

Oops, messed that up :( . I'll upload another one from Monterey later today.

**Comment by williambj1** [comment]
[IntelBT-Monterey.ioreg.zip](https://github.com/headkaze/Hackintool/files/7566946/IntelBT-Monterey.ioreg.zip)

**Comment by headkaze** [comment]
Thanks @williambj1 it all looks good to me. Please check out 3.6.6.

**Comment by telepati** [comment]
Why the same double entry?
<img width="586" alt="Screen Shot 2021-11-19 at 18 21 32" src="https://user-images.githubusercontent.com/3812286/142647446-d2a766d9-10e5-425b-a557-f34937c7eae7.png">

**Comment by headkaze** [comment]
@telepati can you please attach your IOReg file and run the attached version of Hackintool from the command line and post the log output here.
[Hackintool.zip](https://github.com/headkaze/Hackintool/files/7571490/Hackintool.zip)
Specifically I want the output like this:
```
...
2021-11-19 11:11:48.163075-0500 Hackintool[10442:768189] Port: HS14
2021-11-19 11:11:48.163257-0500 Hackintool[10442:768189]    Adding: BCM20702A0
2021-11-19 11:11:48.163349-0500 Hackintool[10442:768189] Port: SS01
...
```
It should at least give me a clue as to why it's adding the device twice.

**Comment by perez987** [comment]
@headkaze I see BT device duplicated also. In IM some users report it three times.
Here's ioreg and log from command line last Hackintool
[GRANJA1.ioreg.zip](https://github.com/headkaze/Hackintool/files/7572091/GRANJA1.ioreg.zip)
[hackintool-command-line.txt](https://github.com/headkaze/Hackintool/files/7572092/hackintool-command-line.txt)

.

**Comment by headkaze** [comment]
Please try 6.6.7. It's now skipping AppleUSBHubPort devices which seem to be where the duplication is occurring.

Only issue I can see with this method is it may not detect a BT dongle in a hub port. If anyone has this issue I'll need to look at their IOReg.

**Comment by telepati** [comment]
@headkaze Here it is.
[IO and Log.zip](https://github.com/headkaze/Hackintool/files/7572598/IO.and.Log.zip)
.

**Comment by perez987** [comment]
Fixed in 3.6.7. Thanks!
